attach: Formalize ROCAttach API

meserve-amd opened the pull request:
## Motivation

Creates a public library and header files for ROCAttach, which is intended to be used by custom tools who wish to utilize the attachment feature for their tool libraries.

Provides both a direct python executable (TODO) and public C library for usage.

I also used this opportunity to do a massive cleanup of the logic in the ptrace session, with the goal of making the library re-entrant as well as easier to understand. More comments will be forthcoming to explain the purpose of each function in PTraceSession.

~~(TODO) finish implementing minor version and error type functions~~

(TODO) more documentation to be added on public ROCAttach library

~~(TODO) more documentation and inline comments to be added on ROCAttach's ptrace session~~

(TODO) Add ptrace theory of operation document
<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Technical Details

Making the API public was the primary goal. This is achieved with public headers and exports. These are modeled off our existing libraries roctx and rocpd.

Re-entrancy was another big goal. detach() now takes a pid to link it to an existing attachment. Also, the static assets in rocattach.cpp were moved into the PTraceSession. Each signal handler thread is now managed by the session mapped to the pid. I also combined the logic so that this thread is the only site where we call waitpid, especially for breakpoints and stop requests during setup.

This also fixes a minor bug where rocprofiler_register::attach() and rocprofiler_register::detach() return values were ignored. They are now correctly piped through.

<!-- Explain the changes along with any relevant GitHub links. -->

## Test Plan

Use existing tests for attach-once and attach-twice

Added test for re-entrancy by having one session attach to multiple running test binaries.

Added test for usage of rocattach from C

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

TBD

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

meserve-amd committed:
attach: Formalize ROCAttach API

- Make ROCAttach public with public headers
- Change detach to take a PID
  - attach and detach are now reentrant
- Cleanup of states and signal handling in ptrace session

meserve-amd commented on projects/rocprofiler-sdk/source/bin/rocprof-attach.py:
This file will be updated extensively once #1365 is merged. This is just to keep the linkage functional for now.

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/rocattach.cpp:
```diff
@@ -140,119 +175,246 @@ build_environment_buffer()
+
+    auto* sessions = CHECK_NOTNULL(get_sessions());
+
+    if(sessions->count(pid) > 0)
```
Can there be multiple callers in here at one time? Multiple attaches? if so, we may need a lock.

meserve-amd replied:
Yes, added

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/ptrace_session.hpp:
```diff
@@ -35,53 +36,116 @@
+        std::atomic<ptrace_session_signal_handler_state_t>& state,
+        std::atomic<rocattach_status_t>&                    error);
+
+    ptrace_session_state_t                             m_state = PTRACE_SESSION_STATE_INITIAL;
```
Does this need to be an atomic?

meserve-amd replied:
state yes, as there are two threads potentially performing read-modify-write operations
error technically no, but since it's cross thread communication I'd rather leave it

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/ptrace_session.cpp:
```diff
@@ -83,203 +122,269 @@ ptrace_op_name(__ptrace_request op)
-: m_pid{_pid}
-{}
-
-PTraceSession::~PTraceSession()
```
Could this destroy while we are waiting in another thread on the pid? We may have some issues if we have another thread waiting on a PID and this is called. Should we wait on pid here as well to make sure its terminated?

meserve-amd replied:
Detaching calls stop_signal_handler which will close out the signal handler gracefully and allow everything to continue. I don't think this particular case is an issue.

meserve-amd committed:
attach: update rocprof-attach

- Fixes mixed up definition of ROCPROF_ATTACH_TOOL_LIBRARY
  - ROCPROF_ATTACH_TOOL_LIBRARY now always means the tool library loaded by the attachment target
  - ROCPROF_ATTACH_LIBRARY refers to the library used to perform attachment
- Add direct call of rocprof-attach
- Fix python library call of rocprof-attach
  - Function now named attach(), changed from main()

meserve-amd committed:
attach: rocprof-compute ROCAttach updates

- Update to new library names
- Correct usage of C lib detach

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/ptrace_session.cpp:
```diff
@@ -65,6 +69,41 @@ typedef struct
+    snprintf(filename, sizeof filename, "/proc/%d/auxv", pid);
+
+    fd = open(filename, O_RDONLY);
+    if(fd < 0) ROCP_ERROR << "[rocprofiler-sdk-rocattach] Unable to open auxv file " << filename;
```
What happens when this fails? Does it need to fatal out? We likely need to handle errors a bit more carefully here.

meserve-amd replied:
This kept the same error style as its original,. This function is slated to be re-written per the other comment, so this is no longer needed.

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/ptrace_session.cpp:
```diff
@@ -327,61 +450,168 @@ PTraceSession::read(size_t addr, std::vector<uint8_t>& data, size_t size) const
+    // Continue until breakpoint is hit
+    ROCATTACH_CALL(cont());
+    while(m_ptrace_signal_handler_state.load() ==
+          PTRACE_SIGNAL_HANDLER_STATE_WAITING_FOR_BREAKPOINT)
```
Is there any case where this could hang? i.e. never hit the breakpoint? We may want to add some sort of retry +  delay scheme here if this is a potential failure point.  At the minimum, lets add a ROCP_INFO/ROCP_TRACE message inside the while loop that triggers every once in a while so we could detect if there was a hang from the logs.

meserve-amd replied:
I agree this needs a retry + backoff delay. Usually this mechanism is pretty reliable, but weird things can happen and the consequences are deadlock.

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/ptrace_session.cpp:
```diff
@@ -48,11 +45,18 @@ static_assert(sizeof(void*) == 8);
+// technique to avoid this is to inject code at the program entry address, as this is extremely
+// unlikely to be called again in a multithreaded process. To determine this address, we inspect the
+// auxv file for the target process.
+/* Following code is copied from glibc's elf.h.  */
```
Is glibc's licensing permissive of this copy? Lets double check its not GPL and is a license type that we can copy from.

meserve-amd replied:
This needs to be re-written.

meserve-amd replied:
Done.

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/rocattach.cpp:
```diff
@@ -140,119 +175,246 @@ build_environment_buffer()
+    status = session.attach();
+    if(status != ROCATTACH_STATUS_SUCCESS)
+    {
+        ROCP_ERROR << "[rocprofiler-sdk-rocattach] Attachment failed to pid " << pid;
```
Lets also print the status here as well.

meserve-amd replied:
done

meserve-amd committed:
address review comments

meserve-amd committed:
attach: add test for rocattach

- Fix some bugs with reentrancy

vedithal-amd commented on projects/rocprofiler-compute/src/rocprof_compute_profile/profiler_rocprofiler_sdk.py:
```diff
@@ -92,7 +92,8 @@ def get_profiler_options(
             options.update({
-                "ROCPROF_ATTACH_TOOL_LIBRARY": rocprofiler_attach_tool_path,
+                "ROCPROF_ATTACH_TOOL_LIBRARY": rocprofiler_sdk_tool_path
+                "ROCPROF_ATTACH_LIBRARY": rocprofiler_attach_library_path,
```
no such variable rocprofiler_attach_library_path

meserve-amd replied:
ah there was a merge conflict here that I resolved and must have missed something, let me fix this.

meserve-amd replied:
Fixed

vedithal-amd commented on projects/rocprofiler-compute/src/rocprof_compute_profile/profiler_rocprofiler_sdk.py:
```diff
@@ -92,7 +92,8 @@ def get_profiler_options(
             )
             options.update({
-                "ROCPROF_ATTACH_TOOL_LIBRARY": rocprofiler_attach_tool_path,
+                "ROCPROF_ATTACH_TOOL_LIBRARY": rocprofiler_sdk_tool_path
```
is ROCPROF_ATTACH_TOOL_LIBRARY still needed if we are using ROCPROF_ATTACH_LIBRARY?

meserve-amd replied:
I opted to make a distinction here since the code previously was not clear on this.
ROCPROF_ATTACH_TOOL_LIBRARY refers to the tool library/libraries that will be used by rocprofiler-sdk inside the target process after attachment.
ROCPROF_ATTACH_LIBRARY refers to the library with attach and detach entrypoints that will be used to perform attachment to the target process. In nearly all cases this is rocattach, which we provide.

ROCPROF_ATTACH_LIBRARY  has sensible defaults and probably doesn't need to be specified.

meserve-amd committed:
address review comments, fix tests, format

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/rocattach.cpp:
```diff
@@ -140,119 +185,273 @@ build_environment_buffer()
-    if(!ptrace_session->call_function("librocprofiler-register.so", "rocprofiler_register_detach"))
+
+    ROCP_TRACE << "[rocprofiler-sdk-rocattach] Attempting detachment to pid " << pid;
+    session->detach();
```
Minor status = session->detach().

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/rocattach.cpp:
```diff
@@ -140,119 +185,273 @@ build_environment_buffer()
+        }
+        return ROCATTACH_STATUS_SUCCESS;
+    }
+    else
```
Can we place an rocp info/warning to state this call was recieved with PID=0 and everything will be detached?

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/rocattach.cpp:
```diff
@@ -140,119 +185,273 @@ build_environment_buffer()
+
+rocattach_status_t
+rocattach_get_version(uint32_t* major, uint32_t* minor, uint32_t* patch)
+{
```
Lets just CHECK() these are non-zero for all of them.

bwelton commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-rocattach/rocattach.cpp:
```diff
@@ -140,119 +185,273 @@ build_environment_buffer()
+rocattach_status_t
+rocattach_get_version_triplet(rocattach_version_triplet_t* info)
+{
+    *info = {.major = ROCATTACH_VERSION_MAJOR,
```
CHECK() this is not null

bwelton reviewed (approved):
Mostly smaller issues. Please resolve and add some docs but otherwise, things are starting to look good. Approving for now.

meserve-amd committed:
attach: add timeouts to ptrace operations

meserve-amd committed:
attach: fix typo in rocprofiler-compute

meserve-amd committed:
Increase timeout for detach output generation

ammarwa committed:
Fixing attach tests from install directory

meserve-amd committed:
attach: disable some sanitizers on attach tests

- Disable ASan, TSan, and UBSan for the new parallel-attach test
- Lower log level for LSan tests, existing behavior from other tests

meserve-amd committed:
WIP add logging for proc maps for RedHat test debug

meserve-amd committed:
revert to earlier test format for output reading

meserve-amd committed:
formattting

itrowbri reviewed (approved):
Approving this since we're looking to merge ASAP. The documentation and signal handling code looks good to me.

vedithal-amd reviewed (approved):
Approved for rocprof-compute changes
